bittide/src/Bittide/Axi4.hs

```diff
@@ -0,0 +1,108 @@
+  , NFDataX userType) =>
+  Circuit
+    (Axi4Stream dom ('Axi4StreamConfig 1 idWidth destWidth) userType)
+    (Axi4Stream dom ('Axi4StreamConfig dataWidth idWidth destWidth) userType)
+axisFromByteStream = Circuit (mealyB go initState)
+ where
+  initState ::
+    (Index dataWidth, Vec (dataWidth - 1) (Unsigned 8, Bool, Bool))
+  initState = (0, repeat initTempAxi)
+  initTempAxi =(0, False, False)
+  go state (Nothing, _) = (state, (Axi4StreamS2M{_tready = True}, Nothing))
+  go (oldCounter, storedBytes) (Just Axi4StreamM2S{..}, Axi4StreamS2M{_tready}) =
+    ((newCounter, nextStoredBytes), (smallS2M, bigM2S))
+   where
+    smallS2M = Axi4StreamS2M{_tready = smallReady}
+    bigM2S
+      | not upscaleDone = Nothing
+      | otherwise       = Just $ Axi4StreamM2S
+        { _tdata = newData
+        , _tkeep = newKeeps
+        , _tstrb = newStrobes
+        , _tuser = _tuser
+        , _tid   = _tid
+        , _tdest = _tdest
+        , _tlast = _tlast}
+    (newData, newKeeps, newStrobes) = unzip3 intermediateBytes
+
+    upscaleDone = oldCounter == maxBound || _tlast
+    smallReady  = not upscaleDone || _tready
+    newCounter
+      | upscaleDone && _tready && _tready = 0
+      | upscaleDone = oldCounter
+      | otherwise   = satSucc SatWrap oldCounter
+
+    intermediateBytes = replace oldCounter (head _tdata, head _tkeep, head _tstrb)
+      (storedBytes :< initTempAxi)
+
+    nextStoredBytes
+      | upscaleDone && _tready = repeat initTempAxi
+      | otherwise              = init intermediateBytes
+
+{-# NOINLINE axisToByteStream #-}
+-- | Transforms an Axi4 stream of n e into an Axi4 stream. of n bytes wide.
+--  It uses an internal counter to select the bytes of the incoming axi stream one by one.
+-- The incoming stream is acknowledged when the last byte is acknowledged by the outgoing stream.
+-- The `_tuser`, _tdest` and `_tid` are ignored and blindly routed to the output.
+axisToByteStream ::
+  forall dom dataWidth idWidth destWidth userType .
+  ( HiddenClockResetEnable dom
+  , KnownNat dataWidth
+  , 1 <= dataWidth
+  , Eq userType
+  , NFDataX userType) =>
+  Circuit
+    (Axi4Stream dom ('Axi4StreamConfig dataWidth idWidth destWidth) userType)
+    (Axi4Stream dom ('Axi4StreamConfig 1 idWidth destWidth) userType)
+axisToByteStream = Circuit (mealyB go (0 :: Index dataWidth))
+ where
+  go state (Nothing, _) = (state, (Axi4StreamS2M{_tready = True}, Nothing))
+  go oldCounter (Just Axi4StreamM2S{..}, Axi4StreamS2M{_tready}) = (newCounter, (bigS2M, smallM2S))
+   where
+    bigS2M = Axi4StreamS2M{_tready = _tready && lastByte}
+    lastByte = oldCounter == lastValidByteIndex
+    smallM2S = Just $ Axi4StreamM2S
+        { _tdata = outData :> Nil
+        , _tkeep = outKeep :> Nil
+        , _tstrb = outStrb :> Nil
+        , _tuser = _tuser
+        , _tid   = _tid
+        , _tdest = _tdest
+        , _tlast = _tlast && lastByte}
+    (outData, outKeep, outStrb) = zip3 _tdata _tkeep _tstrb !! oldCounter
```

Comment:
```suggestion
    (outData, outKeep, outStrb) = zip3 _tdata _tkeep _tstrb !! oldCounter
```

`!!` typically is expensive (i.e., it will become an n-mux), but in this case it cannot be replace by shifting without trading it for reduced throughput. Could you add a comment to the code explaining this?